digdag-cli/src/main/java/io/digdag/cli/profile/TaskAnalyzer.java

```diff
@@ -0,0 +1,97 @@
+package io.digdag.cli.profile;
+
+import com.google.inject.Injector;
+import io.digdag.core.database.TransactionManager;
+import io.digdag.core.session.ArchivedTask;
+import io.digdag.core.session.SessionStoreManager;
+import io.digdag.core.session.StoredSessionAttemptWithSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+public class TaskAnalyzer
+{
+    private static final Logger logger = LoggerFactory.getLogger(TaskAnalyzer.class);
+
+    private final Injector injector;
+
+    public TaskAnalyzer(Injector injector)
+    {
+        this.injector = injector;
+    }
+
+    private void waitAfterDatabaseAccess(int waitMillis)
+    {
+        try {
+            TimeUnit.MILLISECONDS.sleep(waitMillis);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    public TasksSummary run(Instant createdFrom,
+                            Instant createdTo,
+                            int fetchedAttempts,
+                            int partitionSize,
+                            int databaseWaitMillis)
+    {
+        TransactionManager tm = injector.getInstance(TransactionManager.class);
+        SessionStoreManager sm = injector.getInstance(SessionStoreManager.class);
+
+        AtomicLong lastId = new AtomicLong();
+        TasksSummary.Builder tasksSummaryBuilder = new TasksSummary.Builder();
+        logger.debug("Task analysis started");
+
+        while (true) {
+            logger.debug("Collecting attempts");
+
+            // Paginate to avoid too long transaction
+            AtomicLong counter = new AtomicLong();
+            Map<Long, List<StoredSessionAttemptWithSession>> partitionedAttemptIds = tm.begin(() -> {
```

Comment:
I guess the context of this logic is
- Need finished attempts for some duration (createFrom and createTo)
- sm.findFinishedAttemptsWithSessions() return list of attempts
- partitioning the list and set wait to avoid too much stress to DB

IIUC, if sm.findFinishedAttemptsWithSessions return not List but **ResultIterator**, we don't need to do partitioning? (But Digdag not use the interface and I am not sure it is good)
http://jdbi.org/jdbi2/apidocs/org/skife/jdbi/v2/ResultIterator.html

And while running iteration, we can set small wait to avoid stress to DB.
In this case, we can expect to reduce memory consumption of JVM.
In addition, turning parameter 'partition' is a bit difficult to choose proper value and if we don't need it, it may be better.

But I don't have strong opinion, just my thought. 🙂

Reply:
The option sounds cool and simple!

But the description of the class sounds that RDBMS transaction is kept open until the last entry is fetched to me. If so, the long transaction acquires shared locks and it potentially blocks write operation long time. The tuning parameters are actually not easy to understand, but I think few users are interested in them 😉

Reply:
Anyway, the feedback that the implementation of partitioning seems too much is really important. Thank you.

I just updated it https://github.com/treasure-data/digdag/pull/1533/commits/f1f735f35a15d4e1427180e29a76a82bbdf30401